File: java-car-data/car-data-web-ui-tests/src/test/java/AddCarPage.java
import java.util.List;
import java.util.Map;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;


public class AddCarPage {

    private WebDriver driver;

    @FindBy(id = "year")
    private WebElement yearInput;

    @FindBy(id = "make")
    private WebElement makeInput;

    @FindBy(id = "model")
    private WebElement modelInput;

    @FindBy(className = "trimlevel")
    private List<WebElement> trimLevelInputs;

    @FindBy(id = "addtrimlevelbutton")
    private WebElement addTrimLevelButton;

    @FindBy(id = "submitcarbutton")
    private WebElement submitCarButton;

    public AddCarPage(WebDriver driver) {
        this.driver = driver;
        PageFactory.initElements(driver, this);
    }

    public void typeYear(String year) {
        yearInput.sendKeys(year);
    }

    public void typeMake(String make) {
        makeInput.sendKeys(make);
    }

    public void typeModel(String model) {
        modelInput.sendKeys(model);
    }

    public void typeTrimLevels(String trimLevels) {
        String[] trimLevelArray = trimLevels.split("\n");

        for (int i=1; i < trimLevelArray.length; i++) {
            addTrimLevelButton.click();
        }

        for (int i=0; i < trimLevelArray.length; i++) {
            trimLevelInputs.get(i).sendKeys(trimLevelArray[i]);
        }
    }

    public void clickSubmitCarButton() {
        submitCarButton.click();
    }

    public void addCar(Map<String, String> carMap) {
        typeYear(carMap.get("year"));
        typeMake(carMap.get("make"));
        typeModel(carMap.get("model"));
        typeTrimLevels(carMap.get("trimLevels"));

        clickSubmitCarButton();

        WebDriverWait wait = new WebDriverWait(driver, 5);
        wait.until(ExpectedConditions.alertIsPresent());
        driver.switchTo().alert().accept();

        // Wait for javascript redirect
        wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("//div[@class='maincontent']//table[@id='carstable']")));
    }
}

File: python-selenium/pages/add_car_page.py
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.ui import WebDriverWait

from .locators import AddCarPageLocators, MainPageLocators
from .page import BasePage

class AddCarPage(BasePage):
    def __init__(self, driver):
        super().__init__(driver)

        # Wait for submit to be present, others should then also be present
        self.wait = WebDriverWait(self.driver, 10)
        self.submit_car_button = self.wait.until(expected_conditions.presence_of_element_located(AddCarPageLocators.SUBMIT_CAR_BUTTON))

    def add_car(self, car):
        year_input = self.driver.find_element(*AddCarPageLocators.YEAR_INPUT)
        make_input = self.driver.find_element(*AddCarPageLocators.MAKE_INPUT)
        model_input = self.driver.find_element(*AddCarPageLocators.MODEL_INPUT)
        add_trim_level_button = self.driver.find_element(*AddCarPageLocators.ADD_TRIM_LEVEL_BUTTON)

        year_input.send_keys(car['year'])
        make_input.send_keys(car['make'])
        model_input.send_keys(car['model'])

        # Trim Levels
        for i in range(0, len(car['trim_levels'])):
            add_trim_level_button.click()

        trim_levels_inputs = self.driver.find_elements(*AddCarPageLocators.TRIM_LEVEL_INPUTS)

        for i in range(0, len(car['trim_levels'])):
            trim_level = car['trim_levels'][i]
            trim_levels_inputs[i].send_keys(trim_level)

        self.submit_car_button.click()

        # Accept alert
        self.wait.until(expected_conditions.alert_is_present())
        self.driver.switch_to.alert.accept()

        # Wait for javascript redirect
        self.wait.until(expected_conditions.presence_of_element_located(MainPageLocators.CAR_ROWS))